src/main/java/seedu/contax/model/tag/UniqueTagList.java

```diff
@@ -0,0 +1,86 @@
+package seedu.contax.model.tag;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.contax.model.tag.exceptions.DuplicateTagException;
+
+public class UniqueTagList implements Iterable<Tag> {
+    private final ObservableList<Tag> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Tag> internalUnmodifiableList = FXCollections.unmodifiableObservableList(internalList);
+
+    /**
+     * Checks the list if it contains an equivalent tag as the given argument.
+     * @param toCheck The tag to check
+     * @return true if there exists an equivalent tag in the list.
+     */
+    public boolean contains(Tag toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::equals);
+    }
+
+    /**
+     * Adds a tag to the list. The specified tag must not already exist in the list.
+     * @param toAdd The specified tag to add
+     */
+    public void add(Tag toAdd) {
+        requireNonNull(toAdd);
+        if (contains(toAdd)) {
+            throw new DuplicateTagException();
+        }
+        internalList.add(toAdd);
+    }
+
+    public void setTags(UniqueTagList replacement) {
+        requireNonNull(replacement);
+        internalList.setAll(replacement.internalList);
+    }
+
+    public void setTags(List<Tag> tags) {
+        requireNonNull(tags);
+
+        if (!tagsAreUnique(tags)) {
+            throw new DuplicateTagException();
+        }
+
+        internalList.setAll(tags);
+    }
+
+    public ObservableList<Tag> asUnmodifiableObservableList() {
+        return internalUnmodifiableList;
+    }
+
+    @Override
+    public Iterator<Tag> iterator() {
+        return internalList.iterator();
+    }
+
+    private boolean tagsAreUnique(List<Tag> tags) {
+        for (int i = 0; i < tags.size(); i++) {
+            for (int j = i + 1; j < tags.size(); j++) {
+                if (tags.get(i).isSameTag(tags.get(j))) {
+                    return false;
+                }
```

Comment:
maybe can reduce arrow-head a bit
in one line return !tags.get(i).isSameTag(tags.get(j)

Reply:
sorry, my careless, i did not consider that. I found [https://www.tutorialkart.com/java/java-program-find-unique-words-in-a-string/](url) maybe can try using hashset instead of nested loop